### PR TITLE
Improve ServerSideRender Component to retain preview of the component while it is loading new state

### LINKED ISSUE #24519: Improve ServerSideRender UX
The ServerSideRender component is really handy, especially as a gateway for developers to get into building blocks however the constant reloading causes a lot of page resizing and a bit of a clunky experience.

David Gwyer has made a drop in replacement that is a lot nicer and I think it should be merged into core.

You can see the replacement here: https://github.com/dgwyer/server-side-render-x

And here's a gif showing it working:

The code moves the spinner and only replaces the content once it's loaded making the experience feel less clunky.

### PR DESCRIPTION
**Pull request by fabiankaegy**
## Description
The `ServerSideRender` component now continues to show the markup from the previous request when it is fetching the new markup. The reason behind it is preventing the flashing of content when attributes are being changed. This work is a continuation building on the work of @dgwyer here: https://github.com/dgwyer/server-side-render-x

## How has this been tested?

This was tested using the bundled `wp-env` environment with the `Archives` block. 

## Types of changes

- ServerSideRender Component continues to show markup while loading new markup when for example attributes change

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. 
- [x] My code follows the accessibility standards. 
- [x] My code has proper inline documentation. 
- [x] I've included developer documentation if appropriate. 
- [ ] I've updated all React Native files affected by any refactorings/renamings in this PR.

### CODE REVIEW

**Comment by fabiankaegy** [comment]
@dgwyer I'd love to get your thoughts on this first because I used your https://github.com/dgwyer/server-side-render-x repository as an inspiration for the change.

File: packages/server-side-render/src/server-side-render.js
Comment (ZebulanStanphill) [inline packages/server-side-render/src/server-side-render.js:66]: This needs to be a ref at the component level, not a variable local to the current instance of the `fetchData` callback.

File: packages/server-side-render/src/server-side-render.js
Comment (ZebulanStanphill) [inline packages/server-side-render/src/server-side-render.js:77]: This (and several other places) should be checking `isMounted.current` in order to actually read the stored value.

File: packages/server-side-render/src/server-side-render.js
Comment (ZebulanStanphill) [inline packages/server-side-render/src/server-side-render.js:103]: I'm pretty sure the `key` prop here serves no purpose, so this is a good opportunity to remove it.

File: packages/server-side-render/src/server-side-render.js
Comment (ZebulanStanphill) [inline packages/server-side-render/src/server-side-render.js:120]: [`defaultProps` may disappear from React in the future](https://github.com/reactjs/rfcs/pull/107), and it's not necessary for function components anyway since those can just use default arguments. So I'd recommend not using it.

File: packages/server-side-render/src/server-side-render.js
Comment (ZebulanStanphill) [inline packages/server-side-render/src/server-side-render.js:46]: This effect will run on every render because objects in the dependency array are not deep-equality-checked, and so since they are recreated on every render, `props` will always be different because it's technically a new object every time. The only way I know of to avoid re-fetching on every render is to use the `compose` package's `usePrevious` hook and Lodash's `isEqual` function like so:

```suggestion
	const prevProps = usePrevious( props );
	useEffect( () => {
		if ( ! isEqual( prevProps, props ) ) {
			fetchData( props );
		}
	} );
```

File: packages/server-side-render/src/server-side-render.js
Comment (ZebulanStanphill) [inline packages/server-side-render/src/server-side-render.js:48]: Passing in options here isn't really necessary, because this function is already within the scope of the parent component, which means we can just use the props directly.

**Review by ZebulanStanphill — CHANGES_REQUESTED** [review]
Coincidentally, I happened to be working on a PR that just refactors `ServerSideRender` into a function component using React hooks: #28297. (Which your PR also does, in addition to updating the behavior.) I didn't realize that this PR existed until after I'd created mine. :stuck_out_tongue:

It might make more sense for my PR to be reviewed and merged first since it would make this one a lot smaller and easier to review. Either way, feel free to rebase this PR off of mine or just use it as a guide to improve your own. I've managed to spot several issues/potential-tweaks that I've already addressed in my PR.

File: packages/server-side-render/src/server-side-render.js
Comment (fabiankaegy) [reply packages/server-side-render/src/server-side-render.js]: addressed  👍

File: packages/server-side-render/src/server-side-render.js
Comment (fabiankaegy) [reply packages/server-side-render/src/server-side-render.js]: solved  👍

File: packages/server-side-render/src/server-side-render.js
Comment (fabiankaegy) [reply packages/server-side-render/src/server-side-render.js]: Agreed and removed  👍

File: packages/server-side-render/src/server-side-render.js
Comment (fabiankaegy) [reply packages/server-side-render/src/server-side-render.js]: @ZebulanStanphill agreed and done  👍

File: packages/server-side-render/src/server-side-render.js
Comment (fabiankaegy) [reply packages/server-side-render/src/server-side-render.js]: Agreed and done  👍

**Comment by fabiankaegy** [comment]
@ZebulanStanphill Thanks for the review and sorry that you were also working on this at the same time :( 

I'm fine either way. I have addressed all your comments and tested it locally. :) So should be in a better state now 👍

**Review by ZebulanStanphill — DISMISSED** [review]
One of the tests is failing because `package-lock.json` needs to be updated. You can do this by running `npm install`.

File: packages/server-side-render/src/server-side-render.js
Comment (fabiankaegy) [reply packages/server-side-render/src/server-side-render.js]: Ahh you are correct. I have updated my code to also use `useDebounce` in the same way.

File: packages/server-side-render/src/server-side-render.js
Comment (fabiankaegy) [reply packages/server-side-render/src/server-side-render.js]: Solved  👍

File: packages/server-side-render/src/server-side-render.js
Comment (fabiankaegy) [reply packages/server-side-render/src/server-side-render.js]: Yeah you are right  👍 removed  👍

**Comment by fabiankaegy** [comment]
@ZebulanStanphill Thank you for taking the time to review all of this. I have updated all the requested changes :)

File: packages/server-side-render/src/server-side-render.js
Comment (fabiankaegy) [reply packages/server-side-render/src/server-side-render.js]: Thanks for the detailed info. I have made the update accordingly :)

**Comment by Mamaduka** [comment]
Hi, @fabiankaegy

Do you have time to rebase the branch off of the current `trunk`?

**Comment by fabiankaegy** [comment]
@Mamaduka Looking at it now  👍

**Comment by Mamaduka** [comment]
Thanks for the follow-up, @fabiankaegy.

It looks like I missed the notification, but I will try to review this tomorrow.

File: packages/server-side-render/src/server-side-render.js
Comment (Mamaduka) [inline packages/server-side-render/src/server-side-render.js]:
```suggestion
	const prevResponse = usePrevious( response );
```

Just a nitpick: Maybe we should match the naming pattern here.

File: packages/server-side-render/src/server-side-render.js
Comment (Mamaduka) [inline packages/server-side-render/src/server-side-render.js:42]: I think removing the `className` prop will make this a breaking change. We should avoid that.

**Review by Mamaduka — COMMENTED** [review]
Thanks again for working on this, @fabiankaegy.

Changes look great, and I left few minor comments regarding the code.

Here is an example of the updated UX:

https://user-images.githubusercontent.com/240569/131334636-8001065f-bdf9-4fb0-b15b-dedfad58f374.mp4

One thing I'm not sure about is hardcoded inline styles. It will make it harder for developers to override them without providing a custom `LoadingResponsePlaceholder` component. But since the package doesn't ship with styles file, I guess this is our only option.

P.S. Maybe we should match loading state styling to blocks design. Video Block example:

https://user-images.githubusercontent.com/240569/131335086-3f471cf9-2cc1-45ae-a72e-989a4c51ba10.mp4

@jasmussen, any recommendations regarding the styling of the loading state?

**Comment by jasmussen** [comment]
> any recommendations regarding the styling of the loading state?

Thanks for the ping. Overall I think loading states across the board need a holistic rethink. The spinner needs love, and the states themselves could be looked at on whether they need spinners in the first place: if the loading happens fast enough, it's best not to show a spinner at all (i.e. the spinner could appear after a delay).

However the current loading state for the RSS feed is so mediocre that what's shipping already in this branch is an improvement, and since we need to look at loading states holistically, probably best to not hold up this PR for that.

I do share your instinct, though, that if there's prior art, it'd be nice to leverage that, and the prior art for the video is a good one to emulate. I.e. the entire item that is in a loading state becomes semi-transparent, and the spinner is overlaid dead center of the container. Same as for the image:

image

**Comment by Mamaduka** [comment]
Thanks for your input, @jasmussen. It's much appreciated as always.

**Comment by Mamaduka** [comment]
I updated the loading state styling to match existing ones in the core.

@fabiankaegy, @jasmussen, I would love to get your feedback if you think this is in a good state to land.

https://user-images.githubusercontent.com/240569/133982413-4d7d1033-4981-4a8d-a7d7-43b3b661bddb.mp4

File: packages/server-side-render/src/server-side-render.js
Comment (fabiankaegy) [inline packages/server-side-render/src/server-side-render.js]: @Mamaduka The reason why I removed the `className` reference here is because it also is included here: https://github.com/WordPress/gutenberg/pull/28289/files#diff-1fcc38b89edc0d5f7b6239481f42d74b8e2010d792fb26b1e0f8623b222a06fbR160 meaning we now have markup where we nest the className in itself. And I worried that this would lead to stying conflicts. 

Should we maybe do something like `${className}__loading-placeholder` ?

**Comment by fabiankaegy** [comment]
@Mamaduka I think that visual style works and is a great improvement over the current styling because we don't have that huge visual jump. I would be okay with shipping it like that. :) Thank you for your help! :)

File: packages/server-side-render/src/server-side-render.js
Comment (Mamaduka) [reply packages/server-side-render/src/server-side-render.js]: I a good call, and sorry I missed the obvious clue 😄 

The `DefaultLoadingResponsePlaceholder` also received a complete set of props, so if someone is overriding the default placeholder, this change won't break anything.

I'm going to revert the change.

**Comment by jasmussen** [comment]
From a quick glance at the video (thank you for providing!) I see the spinner centered in the box, which in its transient state is semi-transparent. Which seems good to me, thank you!

**Comment by dgwyer** [comment]
Good to see some progress on this. However, I'm not really a fan of the 'dimming' of the block content during the re-render cycle. I can appreciate _why_ it was done this way. i.e. To make it easier to place the spinner in a consistent position central to the block.

Dimming though, introduces unecessary visual cruft for me. Especially for multiple quick fire changes in the block settings panel. I originally created [my own version](https://wpgoplugins.com/an-improved-server-side-render-component-for-dynamic-wordpress-blocks/) of the SSR component to _remove_ as much visual change between the new rendered content and the previous content as possible.

e.g. This is my initial original effort. See how even for rapid settings changes the UI changes look almost as smooth as if it was being updated directly via JS.

**Comment by Mamaduka** [comment]
Thanks for the feedback, @dgwyer.

I tried without dimming, but then Spinner isn't visible in some cases.

After this change, you should be able to override the `LoadingResponsePlaceholder` prop with a custom component and use styling that matches your block's UI. In addition, there will be no need to maintain the custom `ServerSideDrender` component since now you have access to the `prevResponse` as `children` prop.

**Review by Mamaduka — APPROVED** [review]

**Comment by jasmussen** [comment]
Perhaps a separate exploration can be to only show the dimming _and_ the spinner after a delay of a second or so, in an optimistic assumption that the operation will have finished before that.

**Comment by dgwyer** [comment]
@Mamaduka As I mentioned, it's obvious why the dimming was introduced but it doesn't really help solve the original issue. Ideally, no visual changes to the content transition should be introduced apart from the spinner. This was the primary motivation behind the original implementation.

I [already addressed](https://wpgoplugins.com/an-improved-server-side-render-component-for-dynamic-wordpress-blocks/) the issue of the spinner not being visible in some cases by providing a `spinnerLocation` prop to adjust the spinner relative location as required. Perhaps we can implement something similar here rather than introduce unecessary additional UI?

**Comment by dgwyer** [comment]
@jasmussen Debouncing might work. But I think being able to adjust the spinner location should be fine in almost all cases?

**Comment by Mamaduka** [comment]
@dgwyer I would like to avoid controlling individual pieces of the component using props. It could easily increase the complexity of the component.

It's a little more code, but you could achieve the same results by overriding default loading placeholder:
```js
function LoadingResponsePlaceholder( { children } ) {
	return (
		<div style={ { position: 'relative' } }>
			<div
				style={ {
					position: 'absolute',
					top: '10px',
					righ: '0',
				} }
			>
				<Spinner />
			</div>
			{ children }
		</div>
	);
}

<ServerSideRender
	block="core/archives"
	attributes={ attributes }
	LoadingResponsePlaceholder={ LoadingResponsePlaceholder }
/>
```

**Comment by fabiankaegy** [comment]
@Mamaduka Is there anything left that we need to do before this can get merged? Currently Merging is still blocked beause it needs approval.

**Comment by dgwyer** [comment]
@fabiankaegy See my comments above.

**Comment by fabiankaegy** [comment]
@dgwyer I saw your note but thought that it was addressed by the feedback from @Mamaduka here: https://github.com/WordPress/gutenberg/pull/28289#issuecomment-922842458 

Because the `ServerSideRender` component now takes a LoadingPlaceholder component that gives you the ability to customize the experience to your liking this should give the flexibility to achieve the experience you are going after in core.

**Comment by jasmussen** [comment]
> Debouncing might work. But I think being able to adjust the spinner location should be fine in almost all cases?

The postponed spinner feels like it could potentially benefit a lot of these blocks in a generic and reproducible way. The problem with spinner location is that you can't be sure any location is "safe", as it depends on the content inside and how it wraps. Since this PR is an improvement over what's shipping, it feels worth landing it and continuing conversation in new PRs and tickets.

**Comment by dgwyer** [comment]
The point is the **defaut behaviour** shouldn't have any additional layering over existing content during content re-rendering. The 'problem' of spinner location is easily mitigated by displaying it outside the block content. I've been doing this since the beginning and it just works (very well).

Adding UI on top of the content during render transition misses the point entirely. For JS only blocks the content is seamlessly updated, and that's what we want to achieve with SSR blocks too. My solution displays the spinner so it does not intefere with block content at all. See the GIF above (I've since update it so it displays to the left of the block container, vertically centered). This is a much better rendering experience for the user.

I've been using this daily for well over a year and have seen thousands of transitions. Dimming the content and adding the spinner right in the center of the block content during a render transition is not an ideal UX.

**Comment by dgwyer** [comment]
@jasmussen Outide of the block content can be considered a safe location.

I've been using this for a long time and it works very well. As you can see, rapid updating of attributes renders block updates very smoothly. Having the content blinking on/off continuously makes this a very poor UX and is why I implemented this solution in the first place.

https://user-images.githubusercontent.com/1482075/134021853-7773f132-d7cb-403a-a950-61c15ddbd5f7.mp4

**Comment by jasmussen** [comment]
Not necessarily. In the site editor the block might be flush against the top of the screen, which is why we moved the block toolbar below in #29464. Or you might be inserting the block in a narrow context, such as a column, where the block toolbar, or screen edge, will cover the spinner. 

I do appreciate your point, and I do think it's worth following up with a postponed spinner. I think in many cases, certainly those you demo, there reaction time will be so fast there really is no need to see a spinner at all.

**Comment by dgwyer** [comment]
@jasmussen I'll reiterate my point for what it's worth and leave it at that. I would welcome the chance to make this even better for users.

The initial motivation for updating the SSR component was to match as closely the (re)rendering experience of pure JS blocks; which I achieved (with spinner location option). This merged solution does not do that in its current state.

As you pointed out we may not even need the spinner displayed at all in most cases. It would only need to be rendered as a fallback if there is a server delay.

This would be the best all-around solution as it would closely match JS block rendering behavior, fulfilling the original criteria.

**Comment by BinaryMoon** [comment]
Hi - I've been following this thread since I like what @dgwyer had done originally and I am in agreement that no dimming is preferable.

As a user experience it looks so much nicer to not fade out, it feels instant even though it takes the same amount of time. I'm disappointed this (appears to be) merged as is since it doesn't feel any better than what we have currently.

**Comment by jasmussen** [comment]
I created a new ticket, #35027, to improve the transient state so that it only appears if an operation takes a while. I'll see if I can ask a friend to work on it.

**Pull request merged**
